src/lib/consensus/proof_of_stake.ml

```diff
@@ -3916,6 +3914,711 @@ let%test_module "Proof of stake tests" =
+                   always generates pairs of states in short fork range" =
+      let constants = Lazy.force Constants.for_unit_tests in
+      Quickcheck.test (gen_spot_pair_short_misaligned ()) ~f:(fun (a, b) ->
+          assert_consensus_state_pair
+            (With_hash.data a, With_hash.data b)
+            ~assertion:"within short range"
+            ~f:(fun (a, b) -> Hooks.is_short_range a b ~constants) )
+
+    let%test_unit "generator sanity check: gen_spot_pair_long always \
+                   generates pairs of states in long fork range" =
+      let constants = Lazy.force Constants.for_unit_tests in
+      Quickcheck.test gen_spot_pair_long ~f:(fun (a, b) ->
+          assert_consensus_state_pair
+            (With_hash.data a, With_hash.data b)
+            ~assertion:"within long range"
+            ~f:(fun (a, b) -> not (Hooks.is_short_range ~constants a b)) )
+
+    let%test_unit "selection case: equal states" =
+      Quickcheck.test
+        (gen_pair State_hash.gen (gen_spot ()))
+        ~f:(fun (hash, state) ->
+          let hashed_state = {With_hash.data= state; hash} in
+          assert_not_selected (hashed_state, hashed_state) )
+
+    let%test_unit "selection case: aligned checkpoints & different lengths" =
+      Quickcheck.test
+        (gen_spot_pair_short_aligned ~blockchain_length_relativity:`Ascending
+           ())
+        ~f:assert_selected
+
+    let%test_unit "selection case: aligned checkpoints & equal lengths & \
+                   different vrfs" =
+      Quickcheck.test
+        (gen_spot_pair_short_aligned ~blockchain_length_relativity:`Equal
+           ~vrf_output_relativity:`Ascending ())
+        ~f:assert_selected
+
+    let%test_unit "selection case: aligned checkpoints & equal lengths & \
+                   equal vrfs & different hashes" =
+      Quickcheck.test
+        (gen_spot_pair_short_aligned ~blockchain_length_relativity:`Equal
+           ~vrf_output_relativity:`Equal ())
+        ~f:(fun (a, b) ->
+          if State_hash.(With_hash.hash b > With_hash.hash a) then
+            assert_selected (a, b)
+          else assert_selected (b, a) )
+
+    let%test_unit "selection case: misaligned checkpoints & different lengths"
+        =
+      Quickcheck.test
+        (gen_spot_pair_short_misaligned
+           ~blockchain_length_relativity:`Ascending ())
+        ~f:assert_selected
+
+    (* TODO: This test always succeeds, but this could be a false positive as the blockchain length equality constraint
+     * is broken for misaligned short forks.
+     *)
+    let%test_unit "selection case: misaligned checkpoints & equal lengths & \
+                   different vrfs" =
+      Quickcheck.test
+        (gen_spot_pair_short_misaligned ~blockchain_length_relativity:`Equal
+           ~vrf_output_relativity:`Ascending ())
+        ~f:assert_selected
+
+    (* TODO: This test fails because the blockchain length equality constraint is broken for misaligned short forks.
```

Comment:
What's up with this test?

Reply:
The short misaligned spot pair generator needs a little work for this to be possible. The `~blockchain_length_relativity:`Equal` constraint doesn't currently generate equal blockchain lengths for the short misaligned case (it's actually a little tricky to set that up since they need to be in different epochs, but have equal length still).